### PR TITLE
Kill the old ftp storage implementation

### LINKED ISSUE #19941: Drop core's external ftp implementation in favor of icewind1991/files_external_ftp
@icewind1991 did come up with an external storage implementation for ftp based on the new flysystem based concept: https://github.com/icewind1991/files_external_ftp

Proposed steps:
- move https://github.com/icewind1991/files_external_ftp to the owncloud orga
- adjust packaging scripts to pull in the new repo
- drop the old code https://github.com/owncloud/core/blob/master/apps/files_external/lib/ftp.php

@karlitschek @cmonteroluque this would the first step to start working on external file system based on flysystem as discussed for 9.0 - objections? agreement?

I suggest to do this soonish in order to get field experience before addressing other implementations like dropbox etc

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes  #19941

### CODE REVIEW

**Comment by MorrisJobke** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@MorrisJobke hold your horses :wink:

**Comment by MorrisJobke** [comment]
> @MorrisJobke hold your horses

So the new app will be added to this PR too? :)

**Comment by DeepDiver1975** [comment]
> So the new app will be added to this PR too? :)

yes

File: apps/files_external_ftp/appinfo/app.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+$l = \OC::$server->getL10N('files_external_ftp');
+
+OC_Mount_Config::registerBackend('\OCA\Files_External_FTP\FTP', [
```
Comment (Xenopathic) [inline apps/files_external_ftp/appinfo/app.php]: As this is new code going in, can we please use the new backend registration interface?

Comment (DeepDiver1975) [inline apps/files_external_ftp/appinfo/app.php]: sure - push access granted - once travis is green :wink:

**Comment by MorrisJobke** [comment]
```
1) Test\Files_External_FTP\Ftp::testStat
RuntimeException: Could not connect to host: localhost, port:21
```

**Comment by DeepDiver1975** [comment]
travis will be green now - @Xenopathic contribution to adopt the backen registration is more then welcome

@owncloud/qa this change requires migration testing - we need to make sure that mounts still work after migration to the new backend implementation

**Comment by DeepDiver1975** [comment]
@cmonteroluque @karlitschek given the impact I suggest to NOT backport this - THX

**Comment by Xenopathic** [comment]
@DeepDiver1975 Probably best to update the upstream repo directly (cc @icewind1991) then port changes across to here?

**Comment by karlitschek** [comment]
of course not. backports are for critical bugfixes only. not more

File: apps/files_external_ftp/tests/travis/install.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+FTPD=$1
```
Comment (Xenopathic) [inline apps/files_external_ftp/tests/travis/install.sh]: Can we not use the Docker-based testing method for this, as was used in files_external? That way local testing becomes much easier, no worrying about dependencies or other crap like that.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 Probably best to update the upstream repo directly (cc @icewind1991) then port changes across to here?

why - the upstream repo will be discontinued - we did agree on migrating the code into core

File: apps/files_external_ftp/tests/travis/install.sh (hunk 2 above)
Comment (DeepDiver1975) [inline apps/files_external_ftp/tests/travis/install.sh]: No - because you cannot install multiple ftp servers based on the build matrix

**Comment by DeepDiver1975** [comment]
> of course not. backports are for critical bugfixes only. not more

just wanted to make a point -  #19941 is assigned for 9.0.1

File: apps/files_external_ftp/tests/travis/install.sh (hunk 2 above)
Comment (Xenopathic) [inline apps/files_external_ftp/tests/travis/install.sh]: We could use a separate Docker container for each FTP server, the same way we test multiple WebDAV servers.

Comment (DeepDiver1975) [inline apps/files_external_ftp/tests/travis/install.sh]: One could do a lot of things :wink: - let just move on with this here now

File: apps/files_external_ftp/appinfo/app.php (hunk 1 above)
Comment (Xenopathic) [inline apps/files_external_ftp/appinfo/app.php]: The good news is that my term just finished, so I'll almost definitely have time to implement this!

Comment (DeepDiver1975) [inline apps/files_external_ftp/appinfo/app.php]: @Xenopathic feel free to adjust the registration process - THX

**Comment by karlitschek** [comment]
Hmm. 9.0.1 is a bugfix release only. I don't think we should do major changes like that. This is 9.1 from my perspective. 
@cmonteroluque

**Comment by PVince81** [comment]
I agree, I'd also prefer 9.1.
The reason it was set to 9.0.1 was probably because there's a blue ticket attached: https://github.com/owncloud/core/issues/9630#issuecomment-169768967 that could depend on that implementation. :frowning:

**Comment by cmonteroluque** [comment]
ok, agreed

**Comment by DeepDiver1975** [comment]
> @owncloud/qa this change requires migration testing - we need to make sure that mounts still work after migration to the new backend implementation

hello @owncloud/qa :question:

**Comment by DeepDiver1975** [comment]
@Xenopathic THX a lot!

**Comment by PVince81** [comment]
Ticket was moved to 9.1 https://github.com/owncloud/core/issues/19941 (just FYI)

**Comment by SergioBertolinSG** [comment]
I've tested this upgrading a stable9 server with a FTP mount to this branch.

Not working, FTP mount is missing after upgrade.

logs:
```
{"reqId":"Dp81bmo7x1QjomgFmnwb","remoteAddr":"","app":"core","message":"starting upgrade from 9.0.1.3 to 9.1.0.0","level":0,"time":"2016-04-13T08:15:27+00:00","method":"--","url":"--","user":"--"}
{"reqId":"2R8yuMfvtnyAbZ8HX+9m","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:15:59+00:00","method":"GET","url":"\/","user":"admin"}
{"reqId":"65GNlBBUC+RXHHcoWc7X","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:00+00:00","method":"GET","url":"\/index.php\/apps\/files\/","user":"admin"}
{"reqId":"1Txz849zwBrsex01KhLf","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:01+00:00","method":"GET","url":"\/index.php\/core\/js\/oc.js?v=b1280c8a3ec5166b8dae64479012e8ba","user":"admin"}
{"reqId":"z0E6Hb+T87h3lFXjvQCG","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/index.php\/apps\/files_sharing\/api\/externalShares","user":"admin"}
{"reqId":"z3ktk1nE5GV0HDut3hCk","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/","user":"admin"}
{"reqId":"ag95nqLNi4DBl+HwuH5J","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/index.php\/avatar\/admin\/21","user":"admin"}
{"reqId":"pNJX3NpptMPbgHWIKrpO","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/getstoragestats.php?dir=%2F","user":"admin"}
{"reqId":"rQ4ObAwybPv0EcTsnl+3","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/index.php\/avatar\/admin\/21","user":"admin"}
{"reqId":"FE4r5QCQFO1tb2PTNtSc","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/ocs\/v1.php\/apps\/files_external\/api\/v1\/mounts?format=json","user":"admin"}
{"reqId":"FE4r5QCQFO1tb2PTNtSc","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:02+00:00","method":"GET","url":"\/ocs\/v1.php\/apps\/files_external\/api\/v1\/mounts?format=json","user":"admin"}
{"reqId":"bDjblJzQuP5Ogd4ySeIl","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:08+00:00","method":"GET","url":"\/index.php\/settings\/admin","user":"admin"}
{"reqId":"bDjblJzQuP5Ogd4ySeIl","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:08+00:00","method":"GET","url":"\/index.php\/settings\/admin","user":"admin"}
{"reqId":"iW96ikbNhu6M4uFEJCSj","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:09+00:00","method":"GET","url":"\/index.php\/core\/js\/oc.js?v=b1280c8a3ec5166b8dae64479012e8ba","user":"admin"}
{"reqId":"nGIkuyMAmjVbL9HIP\/6M","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:09+00:00","method":"GET","url":"\/index.php\/apps\/files_external\/globalstorages","user":"admin"}
{"reqId":"nGIkuyMAmjVbL9HIP\/6M","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:09+00:00","method":"GET","url":"\/index.php\/apps\/files_external\/globalstorages","user":"admin"}
{"reqId":"XtdidbldiIRH8xsuelXi","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:10+00:00","method":"PROPFIND","url":"\/remote.php\/webdav","user":"admin"}
{"reqId":"eTV6s0vaH5GeZabFo9Jj","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:10+00:00","method":"GET","url":"\/index.php\/settings\/ajax\/checksetup","user":"admin"}
{"reqId":"lGHgOVX9eGuL6WYJbKa5","remoteAddr":"HIDDEN_IP","app":"files_external","message":"Could not load storage: \"Unable to get backend for ftp\"","level":3,"time":"2016-04-13T08:16:10+00:00","method":"GET","url":"\/index.php\/heartbeat","user":"admin"}
```

**Comment by DeepDiver1975** [comment]
Thx

**Comment by icewind1991** [comment]
Looks like a migration issue, the config still refers to the old ftp backend

**Comment by DeepDiver1975** [comment]
> Looks like a migration issue, the config still refers to the old ftp backend

so the old implementation used 'ftp' - right?
what is the key of the new implementation?

can't we simple adopt the new imp to match the old config format?

THX

**Comment by DeepDiver1975** [comment]
> so the old implementation used 'ftp' - right?
> what is the key of the new implementation?
> 
> can't we simple adopt the new imp to match the old config format?
> 
> THX

@icewind1991 @Xenopathic THX

File: apps/files_external_ftp/lib/backend.php [hunk 3]
```diff
@@ -18,33 +18,28 @@
...
+			->setStorageClass('\OCA\Files_External_FTP\FTP')
 			->setText($l->t('FTP'))
 			->addParameters([
 				(new DefinitionParameter('host', $l->t('Host'))),
+				(new DefinitionParameter('port', $l->t('Port')))
+					->setFlag(DefinitionParameter::FLAG_OPTIONAL),
 				(new DefinitionParameter('root', $l->t('Remote subfolder')))
 					->setFlag(DefinitionParameter::FLAG_OPTIONAL),
-				(new DefinitionParameter('secure', $l->t('Secure ftps://')))
+				(new DefinitionParameter('ssl', $l->t('Secure ftps://')))
```
Comment (Xenopathic) [inline apps/files_external_ftp/lib/backend.php]: The storage code looks for 'secure', but you've changed it to 'ssl' here?

**Comment by Xenopathic** [comment]
The new backend still has the same 'ftp' identifier, so old configs should migrate without issue. That error will be logged if the new app isn't being loaded though

**Comment by DeepDiver1975** [comment]
rebased

**Comment by PVince81** [comment]
Ready for review or needs more work ?

**Comment by DeepDiver1975** [comment]
> Ready for review or needs more work ?

not ready - I asked for help on April 15th
https://github.com/owncloud/core/pull/23101#issuecomment-210403745

**Comment by PVince81** [comment]
We're past feature freeze -> 9.2

File: apps/files_external_ftp/appinfo/app.php [hunk 4]
```diff
@@ -1,16 +1,14 @@
 <?php
 
-$l = \OC::$server->getL10N('files_external_ftp');
+\OC::$server->getEventDispatcher()->addListener(
+	'OCA\\Files_External::loadAdditionalBackends', function($event) {
```
Comment (DeepDiver1975) [inline apps/files_external_ftp/appinfo/app.php]: @Xenopathic this is not working due to app loading order. files_external is loaded before files_external_ftp.

What is the plan to fix this?

**Comment by DeepDiver1975** [comment]
obsoleted by #25313 and https://github.com/owncloud/files_external_ftp

**Pull request closed without merging**